crates/wasmparser/src/validator/operators.rs

```diff
@@ -252,62 +252,89 @@ impl OperatorValidator {
     /// simply that something is needed to be popped.
     ///
     /// If `expected` is `Some(T)` then this will be guaranteed to return
-    /// `Some(T)`, and it will only return success if the current block is
+    /// `T`, and it will only return success if the current block is
     /// unreachable or if `T` was found at the top of the operand stack.
     ///
     /// If `expected` is `None` then it indicates that something must be on the
     /// operand stack, but it doesn't matter what's on the operand stack. This
     /// is useful for polymorphic instructions like `select`.
     ///
     /// If `Some(T)` is returned then `T` was popped from the operand stack and
-    /// matches `expected`. If `None` is returned then it means that `None` was
+    /// matches `expected`. If `Bot` is returned then it means that `None` was
     /// expected and the current block is unreachable.
     fn pop_operand(
         &mut self,
         expected: Option<ValType>,
         resources: &impl WasmModuleResources,
-    ) -> OperatorValidatorResult<Option<ValType>> {
+    ) -> OperatorValidatorResult<ValType> {
         let control = self.control.last().unwrap();
-        let actual = if self.operands.len() == control.height {
-            if control.unreachable {
-                None
-            } else {
+        let actual_ty = if self.operands.len() == control.height && control.unreachable {
+            ValType::Bot
+        } else {
+            if self.operands.len() == control.height {
                 let desc = match expected {
                     Some(ty) => ty_to_str(ty),
                     None => "a type".into(),
                 };
                 bail_op_err!("type mismatch: expected {} but nothing on stack", desc)
+            } else {
+                self.operands.pop().unwrap()
             }
-        } else {
-            self.operands.pop().unwrap()
         };
-        if let (Some(actual_ty), Some(expected_ty)) = (actual, expected) {
-            let bad = if self.features.function_references {
-                !resources.matches(actual_ty, expected_ty)
-            } else {
-                actual_ty != expected_ty
-            };
-            if bad {
-                bail_op_err!(
-                    "type mismatch: expected {}, found {}",
-                    ty_to_str(expected_ty),
-                    ty_to_str(actual_ty)
-                )
+        let actual = match expected {
+            None => actual_ty,
+            Some(expected_ty) => {
+                if !resources.matches(actual_ty, expected_ty) {
```

Comment:
Great.   I realized today that <= is equivalent to == for all types besides those added by function references, so this is correct all the time

Reply:
Yep, the `matches` relation conservatively extends the notion of equality over types. Conservative means here: expressions that were true or false before `matches` remain true & false after the introduction of `matches`, i.e. the relation only affects the truth value of the new terms.